design/authorization-handler.md

```diff
@@ -0,0 +1,601 @@
+
+### How to parse authorization data
+
+The authorization data is available in the OIDC request by 2 ways:
+
+* In the scope parameter. The scope parameter is already supported in current Keycloak, but rather in static way
+
+* In the authorization_details parameter as defined in the [RAR specification](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-rar-07).
+
+Once Keycloak parses the data from any of these 2 parameters, it will save them in the ClientAuthorizationContext object. Note that same request can contain both `scope` and 
+`authorization_details` parameter and hence Keycloak should be able to parse both of them in the same request. The idea is to unify the data in the ClientAuthorizationContext under some structure, which will serve as an abstraction layer over the data that was sent via the scope parameter or authorization_details. However Keycloak will still need to have some flag in ClientAuthorizationContext to track how the data was sent. This may be needed for various reasons. 
+
+For example [Grant Management API specification - section 6.4](https://openid.net/specs/fapi-grant-management-01.html#section-6.4) (not yet supported by Keycloak) allows clients to query Keycloak about the granted consents. Note that the returned structure contains both scope and authorization_details and hence it is needed by KC to track how the particular request was sent.
+
+### ClientAuthorizationContext object
+
+Output of the parsing is ClientAuthorizationContext object, which is a list of AuthorizationDetailsEntry objects. Each AuthorizationDetailsEntry contains data like:
+
+* JSON with the authorization details
+* ClientAuthorizationHandler used
+* Source - the parsing method (scope parameter or authorization_details parameter)
+
+#### Java pseudo-structure
+
+```java
+class ClientAuthorizationContext {
+	/**
+ 	*  List of entries of parsed values from "scope" parameter and    
+      * from "authorization_details" .
+ 	*  Each entry represents single parsed value from "scope"
+      * parameter or single entry of "authorization_details"
+ 	*
+ 	* There are also entries corresponding to all "default client
+     * scopes" for the particular client, which are added
+     * automatically.
+ 	*/
+	private List<AuthorizationDetailsEntry> entries;
+}
+
+class AuthorizationDetailsEntry {
+    	// Client scope linked with this authorizationDetails
+    	ClientScopeModel clientScope;
+
+    	// How this entry was added. See enum "Source" below
+    	Source source;
+
+    	// AuthorizationDetails data sent by the client. In case that
+// this was added by "scope" parameter, the "scope" entry is wrapped
+// to the authorizationDetails. See below for the details:
+      AuthorizationDetailsJSONRepresentation authzDetails;  	 
+
+}
+
+enum Source {
```

Comment:
Does the source matters? At the end you are interested on what is being requested, right? Or this is because you want to use this information later to build authorization/token responses?

Reply:
Yes, exactly. In the token response, you may need to know if you include stuff in the `scopes` or `authorization_details` . The Grant Management API also has some differentiation regarding this.

We may not need "Source" in case if we can always know that particular "type" can be addressed either by `scope` parameter or by RAR (`authorization_details` parameter). So for example type "group" will be always handled by scope parameter like `scope=group:123` when the type `payment` will be always covered by RAR. But not sure if such restriction is good as it is a chance that some types might be covered by both `scope` parameter and RAR.

Reply:
The client code (using this API) should know where the information should be included. That is why I was wondering if the source really matters when building responses.

In addition to that, why do we need to treat scopes and authorization details as two different things? Internally, we could probably translate scopes to follow the authorization details (more generic/flexible) semantics and make the API work with a single input.

In your example, `scope=group:123` would be translated to:

```
{
  "type": "http://keycloak.org/auth-type/group",
  "Identifier": "123"
}
```

Reply:
That's exactly the point. It will be ideal if your example can be used and the `group` can be addressed by both the `scope` format and RAR format, which you used in your example. So internally Keycloak SPIs can treat it as same thing and also consent screen can be smart enough to treat those as same thing. We cover this in the design here : https://github.com/dgozalo/keycloak-community/blob/proposal/authorization-handler-clientscopes-enh/design/authorization-handler.md#how-to-parse-authorization-data

Possibly just token response may need to know the "source", so that client "foo", which used scope parameter `scope=group:123` will see it in the `scope` parameter in token response and client "bar", which used RAR will have RAR also in the token response. Also Grant Management API may have some requirement around this...

My point was, that if we don't include "Source" in the request data, it will mean that scope "group" type will need to have hardcoded whether it should be included in the token response in `scope` or in the RAR. Which means that the example you used won't be possible as "group" client scope will have hardcoded whether it can be invoked by the `scope` parameter or by rar.